manifests/jail.pp

```diff
@@ -1,64 +1,101 @@
+#
+# @example Single user
+#   sftp_jail::jail { 'myjail':
+#     user  => 'bbriggs',
+#     group => 'bbriggs',
+#   }
+#
+# @example Shared jail
+#   sftp_jail::jail { 'shared_jail':
+#     user        => 'writeuser',
+#     group       => 'writegroup',
+#     match_group => 'sftpusers',
+#   }
+#
+# @param jail_name
+#   The jails name.
+#
+# @param jail_base
+#   The jail's full path.
+#
+# @param user
+#   The main user that will own the corresponding home directory in the jail,
+#   giving the user a place to land. Also sets user ownership for `/incoming`.
+#
+# @param group
+#   The users group.
+#
+# @param real_home
+#   The real full path users home.
+#
+# @param sub_dirs
+#   This directory structure is enforced in the users Home.
+#
+# @param merge_subdirs
+#   Merge sub_dirs with the default sub_dirs?
+#
+# @param match_group
+#   Set the group that SSHd will look for when redirecting users to the jail.
+#   Useful for shared jails. Defaults to the value of group.
+#
+# @param password_authentication
+#   Can the user login with a password? Public key authentication is generally
+#   recommended and has to be configured outside of the scope of this module.
+#
 define sftp_jail::jail (
-  $jail_name               = $name,
-  $user                    = undef,
-  $group                   = undef,
-  $match_group             = undef,
-  $password_authentication = 'no',
+  Enum['present', 'absent'] $ensure                  = 'present',
+  Sftp_jail::File_name      $jail_name               = $name,
+  Stdlib::Absolutepath      $jail_base               = "${sftp_jail::chroot_base}/${jail_name}",
+  Accounts::User::Name      $user                    = $name,
+  Accounts::User::Name      $group                   = $user,
+  Stdlib::Absolutepath      $real_home               = "${jail_base}/home/${user}",
+  Sftp_jail::Sub_dirs       $sub_dirs                = $sftp_jail::sub_dirs,
+  Boolean                   $merge_subdirs           = false,
+  Accounts::User::Name      $match_group             = $group,
+  Boolean                   $password_authentication = false,
 ) {
   include sftp_jail
-  $jail_base = "${sftp_jail::chroot_base}/${jail_name}"
 
-  if ($match_group) {
-    $ssh_match_group = $match_group
-  }
-  else {
-    $ssh_match_group = $group
-  }
-
-  if !($group) {
-    $group = $user
-  }
-
-  file { $jail_base:
-    ensure  => 'directory',
-    owner   => 'root',
-    group   => 'root',
-    mode    => '0755',
-    require => File[$sftp_jail::chroot_base],
-  }
-
-  file { "${jail_base}/incoming":
-    ensure  => 'directory',
-    owner   => $user,
-    group   => $group,
-    mode    => '0775',
-    require => File[$sftp_jail::chroot_base],
-  }
-
-  file { "${jail_base}/home":
-    ensure  => 'directory',
-    owner   => 'root',
-    group   => 'root',
-    mode    => '0755',
-    require => File[$sftp_jail::chroot_base],
+  unless $ensure == 'absent' {
+    if $ensure == 'absent' {
+      $_ensure = 'absent'
```

Comment:
This seems weird. You're checking unless ensure == 'absent' and if $ensure == 'absent', they can't both be correct.

A better idiom might be something like $directory_ensure = bool2str($ensure == 'present', 'directory', 'absent')

Reply:
This was from #69, I cleaned it up: https://github.com/voxpupuli/puppet-sftp_jail/pull/71/commits/b4d6fd3d27df6aed2fb5dc7c47edfe7d52430757

Reply:
Please have a look at the current implementation. It's cleaner in my eyes. ;)